src/main/java/com/google/step/snippet/servlets/FeedbackServlet.java

```diff
@@ -0,0 +1,174 @@
+    Query queryUser = new Query(USER).setFilter(filterUser);
+    Entity userEntity = datastore.prepare(queryUser).asSingleEntity();
+    if (userEntity == null) {
+      userEntity = new Entity(USER);
+      userEntity.setProperty(ID, id);
+      userEntity.setProperty(EMAIL, userService.getCurrentUser().getEmail());
+    }
+    return userEntity;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    JSONObject json = new JSONObject();
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      String url = request.getParameter(URL);
+      Entity userEntity = getUserEntity(datastore, userService);
+
+      // If a user has previously voted on this card, retrieve their past voting status
+      if (url != null) {
+        ArrayList<String> userUpCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+        ArrayList<String> userDownCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+        if (userUpCards != null && userUpCards.contains(url)) {
+          json.put(STATUS, UPVOTED);
+        } else if (userDownCards != null && userDownCards.contains(url)) {
+          json.put(STATUS, DOWNVOTED);
+        }
+      }
+    }
+    response.setContentType("application/json");
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+    // Invalid POST request if URL is invalid or user is not logged
+    String url = request.getParameter(URL);
+    UserService userService = UserServiceFactory.getUserService();
+    if (url == null || !userService.isUserLoggedIn()) {
+      JSONObject json = new JSONObject();
+      json.put(STATUS, NONE);
+      response.setContentType("application/json");
+      response.getWriter().println(json);
+      return;
+    }
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Entity userEntity = getUserEntity(datastore, userService);
+
+    // Get lists of user's upvoted and downvoted cards
+    ArrayList<String> upCards;
+    if (userEntity.getProperty(UPVOTED) == null) {
+      upCards = new ArrayList<String>();
+    } else {
+      upCards = (ArrayList<String>) userEntity.getProperty(UPVOTED);
+    }
+    ArrayList<String> downCards;
+    if (userEntity.getProperty(DOWNVOTED) == null) {
+      downCards = new ArrayList<String>();
+    } else {
+      downCards = (ArrayList<String>) userEntity.getProperty(DOWNVOTED);
+    }
+
+    // Add or remove card from user's list of upvoted and downvoted cards
+    if (request.getParameter(UP) != null) {
+      if (!upCards.contains(url)) {
+        upCards.add(url);
+        if (downCards.contains(url)) {
+          downCards.remove(url);
```

Comment:
This doesn't seem necessary, as the lack of DOWN as a request parameter already removes it.

In other words:
If UP exists, add it, else remove it.
If DOWN exists, add it, else remove it.